File: mobile/lib/bnpl.js
// bnpl.js: the true cost of an installment or "0% interest" plan (GGives,
// BillEase, Home Credit, Shopee/Lazada installment, card installment). It backs
// out the real effective rate hidden inside a monthly quote and any upfront
// fee, so a "0%" plan that carries a processing fee is unmasked. Pure, offline,
// reuses the loan engine. This is an estimate from the numbers entered, not a
// loan offer. No dashes.

import { effectiveMonthlyRate, effectiveAnnualRate } from './loan';

const num = (x) => (Number.isFinite(Number(x)) ? Number(x) : 0);

// fields: { cashPrice, downpayment, months, monthlyPayment, upfrontFee }
// Returns the totals, how much more than paying cash it costs, and the real
// effective monthly and annual rate on the credit actually received.
export function bnplCost(fields = {}) {
  const f = fields || {};
  const cash = Math.max(0, num(f.cashPrice));
  const down = Math.min(Math.max(0, num(f.downpayment)), cash);
  const fee = Math.max(0, num(f.upfrontFee));
  const months = Math.min(Math.max(1, Math.round(num(f.months)) || 1), 60);
  const monthly = Math.max(0, num(f.monthlyPayment));

  // What you finance is the price left after any downpayment. What you pay is
  // the downpayment now, the fee now, and the monthly installments over time.
  const financed = Math.max(0, cash - down);
  const totalPaid = down + fee + monthly * months;
  const extraCost = Math.max(0, totalPaid - cash); // more than paying cash today

  // The numbers do not add up when what you pay in total is less than the cash
  // price: the installments do not even cover the item. A true cost tool must
  // never reassure someone on impossible numbers, so this is its own state.
  const underpays = totalPaid < cash - 0.005;

  // The honest effective rate is measured on the credit actually received: the
  // financed amount minus any fee taken upfront. So a fee shows up as real cost
  // even when the quoted rate is "0%".
  const netCredit = Math.max(0, financed - fee);
  const monthlyRate = effectiveMonthlyRate(netCredit, monthly, months);
  const annualRate = effectiveAnnualRate(monthlyRate);
  // The rate is only meaningful when there is positive net credit to charge it
  // on. When a fee meets or exceeds what is financed, the rate math zeroes out,
  // so fall back to the plain peso extra-cost framing instead of "0%".
  const rateReliable = netCredit > 0.005 && monthlyRate > 0;

  // Genuinely free only when it fully covers the price, costs nothing over
  // cash, and carries no fee.
  const trulyFree = !underpays && extraCost <= 0.005 && fee <= 0.005;

  return {
    cash,
    down,
    fee,
    months,
    monthly,
    financed,
    netCredit,
    totalPaid,
    extraCost,
    monthlyRate,
    annualRate,
    rateReliable,
    underpays,
    trulyFree,
  };
}

File: flutter/lib/money/bnpl.dart
// The true cost of an installment or "0% interest" plan (GGives, BillEase,
// Home Credit, Shopee and Lazada installment, card installment), ported 1:1
// from mobile/lib/bnpl.js. Backs out the real effective rate hidden inside
// a monthly quote and any upfront fee, so a "0%" plan with a processing fee
// is unmasked. Reuses the golden-locked loan engine. An estimate from the
// numbers entered, not a loan offer. Golden-verified against the real RN
// module.

import 'ledger.dart' show amountOf;
import 'loan.dart' show effectiveAnnualRate, effectiveMonthlyRate;

double _jsRound(num x) => (x + 0.5).floorToDouble();

/// fields: cashPrice, downpayment, months, monthlyPayment, upfrontFee.
/// Returns the totals, how much more than paying cash it costs, and the
/// real effective monthly and annual rate on the credit actually received.
Map<String, dynamic> bnplCost([Map<String, dynamic>? fields]) {
  final f = fields ?? const {};
  final cash = amountOf(f['cashPrice']) > 0 ? amountOf(f['cashPrice']) : 0.0;
  final downRaw =
      amountOf(f['downpayment']) > 0 ? amountOf(f['downpayment']) : 0.0;
  final down = downRaw < cash ? downRaw : cash;
  final fee = amountOf(f['upfrontFee']) > 0 ? amountOf(f['upfrontFee']) : 0.0;
  final monthsRounded = _jsRound(amountOf(f['months']));
  final monthsFloor = monthsRounded > 1 ? monthsRounded : 1.0;
  final months = (monthsFloor < 60 ? monthsFloor : 60.0).toInt();
  final monthly =
      amountOf(f['monthlyPayment']) > 0 ? amountOf(f['monthlyPayment']) : 0.0;

  // What you finance is the price left after any downpayment. What you pay
  // is the downpayment now, the fee now, and the installments over time.
  final financedRaw = cash - down;
  final financed = financedRaw > 0 ? financedRaw : 0.0;
  final totalPaid = down + fee + monthly * months;
  final extraRaw = totalPaid - cash;
  final extraCost = extraRaw > 0 ? extraRaw : 0.0;

  // The numbers do not add up when the payments do not even cover the item;
  // a true cost tool must never reassure someone on impossible numbers.
  final underpays = totalPaid < cash - 0.005;

  // The honest effective rate is measured on the credit actually received:
  // financed minus any fee taken upfront, so a fee shows as real cost even
  // when the quoted rate is "0%".
  final netRaw = financed - fee;
  final netCredit = netRaw > 0 ? netRaw : 0.0;
  final monthlyRate = effectiveMonthlyRate(netCredit, monthly, months);
  final annualRate = effectiveAnnualRate(monthlyRate);
  final rateReliable = netCredit > 0.005 && monthlyRate > 0;

  // Genuinely free only when it fully covers the price, costs nothing over
  // cash, and carries no fee.
  final trulyFree = !underpays && extraCost <= 0.005 && fee <= 0.005;

  return {
    'cash': cash,
    'down': down,
    'fee': fee,
    'months': months,
    'monthly': monthly,
    'financed': financed,
    'netCredit': netCredit,
    'totalPaid': totalPaid,
    'extraCost': extraCost,
    'monthlyRate': monthlyRate,
    'annualRate': annualRate,
    'rateReliable': rateReliable,
    'underpays': underpays,
    'trulyFree': trulyFree,
  };
}
